File: cpp/shared/chess_engine.cpp
#include "chess_engine.h"
using namespace chess;

// ── Internal helpers ──────────────────────────────────────────────────────────
static bool inb(int r,int c){ return r>=0&&r<8&&c>=0&&c<8; }

static Board applyRaw(const Board& board, const Move& m) {
    Board b = board;
    int piece = b[m.r][m.c];
    b[m.tr][m.tc] = m.promo.value_or(piece);
    b[m.r][m.c]   = 0;
    if (m.ep) b[m.r][m.tc] = 0;
    switch (m.castle) {
        case 'K': b[7][5]=R;  b[7][7]=0; break;
        case 'Q': b[7][3]=R;  b[7][0]=0; break;
        case 'k': b[0][5]=-R; b[0][7]=0; break;
        case 'q': b[0][3]=-R; b[0][0]=0; break;
    }
    return b;
}

static bool attacked(const Board& b, int r, int c, int by) {
    const int dirs4[4][2]={{1,0},{-1,0},{0,1},{0,-1}};
    for (auto& d : dirs4) {
        int tr=r+d[0],tc=c+d[1];
        while(inb(tr,tc)){int p=b[tr][tc];if(!p){tr+=d[0];tc+=d[1];continue;}if(p==R*by||p==Q*by){return true;}break;}
    }
    const int diag[4][2]={{1,1},{1,-1},{-1,1},{-1,-1}};
    for (auto& d : diag) {
        int tr=r+d[0],tc=c+d[1];
        while(inb(tr,tc)){int p=b[tr][tc];if(!p){tr+=d[0];tc+=d[1];continue;}if(p==B*by||p==Q*by){return true;}break;}
    }
    const int kn[8][2]={{-2,-1},{-2,1},{-1,-2},{-1,2},{1,-2},{1,2},{2,-1},{2,1}};
    for (auto& d : kn) { if(inb(r+d[0],c+d[1])&&b[r+d[0]][c+d[1]]==N*by) { return true; } }
    int pd = by==1?1:-1;
    if(inb(r+pd,c-1)&&b[r+pd][c-1]==P*by) { return true; }
    if(inb(r+pd,c+1)&&b[r+pd][c+1]==P*by) { return true; }
    for(int dr=-1;dr<=1;dr++) { for(int dc=-1;dc<=1;dc++){
        if(!dr&&!dc) { continue; }
        if(inb(r+dr,c+dc)&&b[r+dr][c+dc]==K*by) { return true; }
    }}
    return false;
}

static bool inCheck(const Board& b, int turn) {
    for(int r=0;r<8;r++) { for(int c=0;c<8;c++) { if(b[r][c]==K*turn) { return attacked(b,r,c,-turn); } } }
    return false;
}

static void pawnMoves(const Board& b, int r, int c, int turn, int epR, int epC, std::vector<Move>& out) {
    int d=-turn, start=turn==1?6:1, promo=turn==1?0:7;
    if(inb(r+d,c)&&!b[r+d][c]) {
        if(r+d==promo){for(int q:{Q*turn,R*turn,B*turn,N*turn}) { out.push_back({r,c,r+d,c,q,0,false,false}); }}
        else{out.push_back({r,c,r+d,c});if(r==start&&!b[r+2*d][c]) { out.push_back({r,c,r+2*d,c,{},0,false,true}); }}
    }
    for(int dc:{-1,1}){
        int tr=r+d,tc2=c+dc; if(!inb(tr,tc2)) continue;
        bool cap=b[tr][tc2]&&(b[tr][tc2]>0)!=(turn>0), ep=tr==epR&&tc2==epC;
        if(cap||ep){
            if(tr==promo){for(int q:{Q*turn,R*turn,B*turn,N*turn}) { out.push_back({r,c,tr,tc2,q,0,ep,false}); }}
            else { out.push_back({r,c,tr,tc2,{},0,ep,false}); }
        }
    }
}

static void sliderMoves(const Board& b, int r, int c, int turn, const int dirs[][2], int nd, std::vector<Move>& out) {
    for(int i=0;i<nd;i++){
        int tr=r+dirs[i][0],tc=c+dirs[i][1];
        while(inb(tr,tc)){
            if(b[tr][tc]){if((b[tr][tc]>0)!=(turn>0)){out.push_back({r,c,tr,tc});}break;}
            out.push_back({r,c,tr,tc}); tr+=dirs[i][0]; tc+=dirs[i][1];
        }
    }
}

static std::vector<Move> pseudoMoves(const Board& b, int r, int c, int turn,
                                      bool wK,bool wQ,bool bK,bool bQ,int epR,int epC) {
    std::vector<Move> out;
    int p=std::abs(b[r][c]);
    if(p==P){ pawnMoves(b,r,c,turn,epR,epC,out); }
    else if(p==N){
        const int d[8][2]={{-2,-1},{-2,1},{-1,-2},{-1,2},{1,-2},{1,2},{2,-1},{2,1}};
        for(auto& dd:d){ int tr=r+dd[0],tc=c+dd[1]; if(inb(tr,tc)&&!(b[tr][tc]&&(b[tr][tc]>0)==(turn>0))) { out.push_back({r,c,tr,tc}); } }
    } else if(p==B){ const int d[4][2]={{1,1},{1,-1},{-1,1},{-1,-1}};sliderMoves(b,r,c,turn,d,4,out); }
      else if(p==R){ const int d[4][2]={{1,0},{-1,0},{0,1},{0,-1}};sliderMoves(b,r,c,turn,d,4,out); }
      else if(p==Q){ const int d1[4][2]={{1,1},{1,-1},{-1,1},{-1,-1}};sliderMoves(b,r,c,turn,d1,4,out);
                     const int d2[4][2]={{1,0},{-1,0},{0,1},{0,-1}};sliderMoves(b,r,c,turn,d2,4,out); }
    else if(p==K){
        for(int dr=-1;dr<=1;dr++) { for(int dc=-1;dc<=1;dc++){
            if(!dr&&!dc) { continue; } int tr=r+dr,tc2=c+dc;
            if(inb(tr,tc2)&&!(b[tr][tc2]&&(b[tr][tc2]>0)==(turn>0))) { out.push_back({r,c,tr,tc2}); }
        }}
        if(turn==1&&r==7&&c==4){
            if(wK&&!b[7][5]&&!b[7][6]&&!attacked(b,7,4,-1)&&!attacked(b,7,5,-1)) { out.push_back({7,4,7,6,{},'K'}); }
            if(wQ&&!b[7][3]&&!b[7][2]&&!b[7][1]&&!attacked(b,7,4,-1)&&!attacked(b,7,3,-1)) { out.push_back({7,4,7,2,{},'Q'}); }
        }
        if(turn==-1&&r==0&&c==4){
            if(bK&&!b[0][5]&&!b[0][6]&&!attacked(b,0,4,1)&&!attacked(b,0,5,1)) { out.push_back({0,4,0,6,{},'k'}); }
            if(bQ&&!b[0][3]&&!b[0][2]&&!b[0][1]&&!attacked(b,0,4,1)&&!attacked(b,0,3,1)) { out.push_back({0,4,0,2,{},'q'}); }
        }
    }
    return out;
}

std::vector<Move> chess::getLegalMoves(const Board& board,int turn,bool wK,bool wQ,bool bK,bool bQ,int epR,int epC){
    std::vector<Move> legal;
    for(int r=0;r<8;r++) { for(int c=0;c<8;c++){
        int p=board[r][c]; if(!p||(p>0)!=(turn>0)) { continue; }
        for(const auto& m:pseudoMoves(board,r,c,turn,wK,wQ,bK,bQ,epR,epC)){
            Board nb=applyRaw(board,m); if(!inCheck(nb,turn)) { legal.push_back(m); }
        }
    }}
    return legal;
}

Game chess::newGame(){
    Game g; const int back[8]={R,N,B,Q,K,B,N,R};
    for(int c=0;c<8;c++) { g.board[0][c]=-back[c]; g.board[1][c]=-P; g.board[6][c]=P; g.board[7][c]=back[c]; }
    g.legalMoves=getLegalMoves(g.board,1,true,true,true,true,-1,-1); return g;
}

Game chess::applyMove(const Game& g, const Move& m){
    Game ng=g; ng.board=applyRaw(g.board,m);
    int piece=std::abs(g.board[m.r][m.c]);
    if(piece==K){if(g.turn==1){ng.wK=ng.wQ=false;}else{ng.bK=ng.bQ=false;}}
    if(piece==R){if(m.r==7&&m.c==7){ng.wK=false;}if(m.r==7&&m.c==0){ng.wQ=false;}if(m.r==0&&m.c==7){ng.bK=false;}if(m.r==0&&m.c==0){ng.bQ=false;}}
    ng.epR=-1;ng.epC=-1;if(m.doublePush){ng.epR=(m.r+m.tr)/2;ng.epC=m.c;}
    ng.turn=-g.turn;
    ng.legalMoves=getLegalMoves(ng.board,ng.turn,ng.wK,ng.wQ,ng.bK,ng.bQ,ng.epR,ng.epC);
    bool chk=inCheck(ng.board,ng.turn);
    ng.status=ng.legalMoves.empty()?(chk?Status::Checkmate:Status::Stalemate):(chk?Status::Check:Status::Active);
    ng.halfMove=(piece==P||ng.board[m.tr][m.tc]!=0)?0:g.halfMove+1;
    ng.fullMove=g.fullMove+(g.turn==-1?1:0);
    return ng;
}

static int evaluate(const Board& b){
    int score=0;
    for(int r=0;r<8;r++) { for(int c=0;c<8;c++){
        int p=b[r][c]; if(!p) { continue; }
        int s=p>0?1:-1, a=std::abs(p), idx=p>0?r*8+c:(7-r)*8+c;
        score+=s*(pieceVal(p)+pstFor(a)[idx]);
    }}
    return score;
}

static int mvvLva(const Board& b,const Move& m){
    int v=std::abs(b[m.tr][m.tc]),att=std::abs(b[m.r][m.c]);
    return v?pieceVal(v)*10-pieceVal(att):0;
}

static int minimax(const Game& g,int depth,int alpha,int beta,int maximising){
    if(depth==0||g.status==Status::Checkmate||g.status==Status::Stalemate) { return evaluate(g.board)*maximising; }
    auto moves=g.legalMoves;
    std::sort(moves.begin(),moves.end(),[&](const Move& a,const Move& b){return mvvLva(g.board,a)>mvvLva(g.board,b);});
    int best=INT_MIN/2;
    for(const auto& m:moves){
        int score=-minimax(applyMove(g,m),depth-1,-beta,-alpha,-maximising);
        best=std::max(best,score); alpha=std::max(alpha,score); if(alpha>=beta) { break; }
    }
    return best;
}

std::optional<Move> chess::getAiMove(const Game& g){
    if(g.legalMoves.empty()) return std::nullopt;
    const int depths[]={0,1,1,2,2,2,3,3,4,4};
    int depth=depths[std::min((int)g.legalMoves.size(),9)];
    int bestScore=INT_MIN/2; Move bestMove=g.legalMoves[0];
    for(const auto& m:g.legalMoves){
        int score=-minimax(applyMove(g,m),depth,INT_MIN/2,INT_MAX/2,-g.turn);
        if(score>bestScore) { bestScore=score; bestMove=m; }
    }
    return bestMove;
}

File: dart/shared/lib/chess_engine.dart
/// Chess engine — Dart. Sound null safety, value-like immutable moves.
library chess_engine;

const int P=1,N=2,B=3,R=4,Q=5,K=6;

enum ChessStatus { active, check, checkmate, stalemate }

class ChessMove {
  final int r,c,tr,tc;
  final int? promo;
  final String? castle;
  final bool ep, doublePush;
  const ChessMove(this.r,this.c,this.tr,this.tc,{this.promo,this.castle,this.ep=false,this.doublePush=false});
  @override bool operator==(Object o) => o is ChessMove && r==o.r&&c==o.c&&tr==o.tr&&tc==o.tc&&promo==o.promo;
  @override int get hashCode => Object.hash(r,c,tr,tc,promo);
}

class ChessGame {
  final List<List<int>> board;
  final int turn;
  final bool wK,wQ,bK,bQ;
  final int epR,epC,halfMove,fullMove;
  final ChessStatus status;
  final List<ChessMove> legalMoves;
  const ChessGame({required this.board,required this.turn,required this.wK,required this.wQ,
    required this.bK,required this.bQ,required this.epR,required this.epC,
    required this.halfMove,required this.fullMove,required this.status,required this.legalMoves});
}

// ── PST tables ───────────────────────────────────────────────────────────────
const _PST_P = [ 0,0,0,0,0,0,0,0, 50,50,50,50,50,50,50,50, 10,10,20,30,30,20,10,10,
   5,5,10,25,25,10,5,5, 0,0,0,20,20,0,0,0, 5,-5,-10,0,0,-10,-5,5, 5,10,10,-20,-20,10,10,5, 0,0,0,0,0,0,0,0];
const _PST_N = [-50,-40,-30,-30,-30,-30,-40,-50,-40,-20,0,0,0,0,-20,-40,-30,0,10,15,15,10,0,-30,
  -30,5,15,20,20,15,5,-30,-30,0,15,20,20,15,0,-30,-30,5,10,15,15,10,5,-30,-40,-20,0,5,5,0,-20,-40,-50,-40,-30,-30,-30,-30,-40,-50];
const _PST_B = [-20,-10,-10,-10,-10,-10,-10,-20,-10,0,0,0,0,0,0,-10,-10,0,5,10,10,5,0,-10,
  -10,5,5,10,10,5,5,-10,-10,0,10,10,10,10,0,-10,-10,10,10,10,10,10,10,-10,-10,5,0,0,0,0,5,-10,-20,-10,-10,-10,-10,-10,-10,-20];
const _PST_R = [0,0,0,0,0,0,0,0, 5,10,10,10,10,10,10,5,-5,0,0,0,0,0,0,-5,-5,0,0,0,0,0,0,-5,-5,0,0,0,0,0,0,-5,-5,0,0,0,0,0,0,-5,-5,0,0,0,0,0,0,-5,0,0,0,5,5,0,0,0];
const _PST_Q = [-20,-10,-10,-5,-5,-10,-10,-20,-10,0,0,0,0,0,0,-10,-10,0,5,5,5,5,0,-10,-5,0,5,5,5,5,0,-5,0,0,5,5,5,5,0,-5,-10,5,5,5,5,5,0,-10,-10,0,5,0,0,0,0,-10,-20,-10,-10,-5,-5,-10,-10,-20];
const _PST_K = [-30,-40,-40,-50,-50,-40,-40,-30,-30,-40,-40,-50,-50,-40,-40,-30,-30,-40,-40,-50,-50,-40,-40,-30,-30,-40,-40,-50,-50,-40,-40,-30,-20,-30,-30,-40,-40,-30,-30,-20,-10,-20,-20,-20,-20,-20,-20,-10,20,20,0,0,0,0,20,20,20,30,10,0,0,10,30,20];
final _PST = {P:_PST_P,N:_PST_N,B:_PST_B,R:_PST_R,Q:_PST_Q,K:_PST_K};
final _VAL = {P:100,N:320,B:330,R:500,Q:900,K:20000};

// ── Helpers ──────────────────────────────────────────────────────────────────
bool _inb(int r,int c) => r>=0&&r<8&&c>=0&&c<8;

List<List<int>> _cloneBoard(List<List<int>> b) => [for(var row in b) [...row]];

List<List<int>> _applyRaw(List<List<int>> board, ChessMove m) {
  final b = _cloneBoard(board);
  final piece = b[m.r][m.c];
  b[m.tr][m.tc] = m.promo ?? piece; b[m.r][m.c] = 0;
  if (m.ep) b[m.r][m.tc] = 0;
  switch (m.castle) {
    case 'K': b[7][5]=R;  b[7][7]=0;
    case 'Q': b[7][3]=R;  b[7][0]=0;
    case 'k': b[0][5]=-R; b[0][7]=0;
    case 'q': b[0][3]=-R; b[0][0]=0;
  }
  return b;
}

bool _attacked(List<List<int>> b, int r, int c, int by) {
  for (final d in [[1,0],[-1,0],[0,1],[0,-1]]) {
    var tr=r+d[0],tc=c+d[1];
    while(_inb(tr,tc)){ final p=b[tr][tc]; if(p==0){tr+=d[0];tc+=d[1];continue;} if(p==R*by||p==Q*by)return true; break; }
  }
  for (final d in [[1,1],[1,-1],[-1,1],[-1,-1]]) {
    var tr=r+d[0],tc=c+d[1];
    while(_inb(tr,tc)){ final p=b[tr][tc]; if(p==0){tr+=d[0];tc+=d[1];continue;} if(p==B*by||p==Q*by)return true; break; }
  }
  for (final d in [[-2,-1],[-2,1],[-1,-2],[-1,2],[1,-2],[1,2],[2,-1],[2,1]])
    if(_inb(r+d[0],c+d[1])&&b[r+d[0]][c+d[1]]==N*by) return true;
  final pd=by==1?1:-1;
  if(_inb(r+pd,c-1)&&b[r+pd][c-1]==P*by) return true;
  if(_inb(r+pd,c+1)&&b[r+pd][c+1]==P*by) return true;
  for(var dr=-1;dr<=1;dr++) for(var dc=-1;dc<=1;dc++){
    if(dr==0&&dc==0) continue;
    if(_inb(r+dr,c+dc)&&b[r+dr][c+dc]==K*by) return true;
  }
  return false;
}

bool _inCheck(List<List<int>> b, int turn) {
  for(var r=0;r<8;r++) for(var c=0;c<8;c++) if(b[r][c]==K*turn) return _attacked(b,r,c,-turn);
  return false;
}

// ── Move generation ──────────────────────────────────────────────────────────
Iterable<ChessMove> _pseudoMoves(List<List<int>> b, int r, int c, int turn,
    bool wK, bool wQ, bool bK, bool bQ, int epR, int epC) sync* {
  final p = b[r][c].abs();
  if (p==P) {
    final d=-turn, start=turn==1?6:1, promo=turn==1?0:7;
    if (_inb(r+d,c)&&b[r+d][c]==0) {
      if (r+d==promo) { for (final q in [Q*turn,R*turn,B*turn,N*turn]) yield ChessMove(r,c,r+d,c,promo:q); }
      else { yield ChessMove(r,c,r+d,c); if(r==start&&b[r+2*d][c]==0) yield ChessMove(r,c,r+2*d,c,doublePush:true); }
    }
    for (final dc in [-1,1]) {
      final tr=r+d,tc=c+dc; if(!_inb(tr,tc)) continue;
      final cap=b[tr][tc]!=0&&(b[tr][tc]>0)!=(turn>0), ep=tr==epR&&tc==epC;
      if (cap||ep) {
        if (tr==promo) { for (final q in [Q*turn,R*turn,B*turn,N*turn]) yield ChessMove(r,c,tr,tc,promo:q,ep:ep); }
        else yield ChessMove(r,c,tr,tc,ep:ep);
      }
    }
  } else if (p==N) {
    for (final d in [[-2,-1],[-2,1],[-1,-2],[-1,2],[1,-2],[1,2],[2,-1],[2,1]]) {
      final tr=r+d[0],tc=c+d[1];
      if(_inb(tr,tc)&&!(b[tr][tc]!=0&&(b[tr][tc]>0)==(turn>0))) yield ChessMove(r,c,tr,tc);
    }
  } else if (p==B||p==R||p==Q) {
    final dirs = p==B?[[1,1],[1,-1],[-1,1],[-1,-1]]:p==R?[[1,0],[-1,0],[0,1],[0,-1]]:[[1,1],[1,-1],[-1,1],[-1,-1],[1,0],[-1,0],[0,1],[0,-1]];
    for (final d in dirs) {
      var tr=r+d[0],tc=c+d[1];
      while(_inb(tr,tc)) {
        if(b[tr][tc]!=0){ if((b[tr][tc]>0)!=(turn>0)) yield ChessMove(r,c,tr,tc); break; }
        yield ChessMove(r,c,tr,tc); tr+=d[0]; tc+=d[1];
      }
    }
  } else if (p==K) {
    for(var dr=-1;dr<=1;dr++) for(var dc=-1;dc<=1;dc++){
      if(dr==0&&dc==0) continue;
      final tr=r+dr,tc=c+dc;
      if(_inb(tr,tc)&&!(b[tr][tc]!=0&&(b[tr][tc]>0)==(turn>0))) yield ChessMove(r,c,tr,tc);
    }
    if(turn==1&&r==7&&c==4) {
      if(wK&&b[7][5]==0&&b[7][6]==0&&!_attacked(b,7,4,-1)&&!_attacked(b,7,5,-1)) yield ChessMove(7,4,7,6,castle:'K');
      if(wQ&&b[7][3]==0&&b[7][2]==0&&b[7][1]==0&&!_attacked(b,7,4,-1)&&!_attacked(b,7,3,-1)) yield ChessMove(7,4,7,2,castle:'Q');
    }
    if(turn==-1&&r==0&&c==4) {
      if(bK&&b[0][5]==0&&b[0][6]==0&&!_attacked(b,0,4,1)&&!_attacked(b,0,5,1)) yield ChessMove(0,4,0,6,castle:'k');
      if(bQ&&b[0][3]==0&&b[0][2]==0&&b[0][1]==0&&!_attacked(b,0,4,1)&&!_attacked(b,0,3,1)) yield ChessMove(0,4,0,2,castle:'q');
    }
  }
}

// ── Public API ───────────────────────────────────────────────────────────────
List<ChessMove> getLegalMoves(List<List<int>> board, int turn,
    bool wK, bool wQ, bool bK, bool bQ, int epR, int epC) {
  final legal = <ChessMove>[];
  for(var r=0;r<8;r++) for(var c=0;c<8;c++) {
    if(board[r][c]!=0&&(board[r][c]>0)==(turn>0))
      for(final m in _pseudoMoves(board,r,c,turn,wK,wQ,bK,bQ,epR,epC))
        if(!_inCheck(_applyRaw(board,m),turn)) legal.add(m);
  }
  return legal;
}

ChessGame newGame() {
  final board = List.generate(8,(_)=>List.filled(8,0));
  final back=[R,N,B,Q,K,B,N,R];
  for(var c=0;c<8;c++){board[0][c]=-back[c];board[1][c]=-P;board[6][c]=P;board[7][c]=back[c];}
  final moves=getLegalMoves(board,1,true,true,true,true,-1,-1);
  return ChessGame(board:board,turn:1,wK:true,wQ:true,bK:true,bQ:true,epR:-1,epC:-1,halfMove:0,fullMove:1,status:ChessStatus.active,legalMoves:moves);
}

ChessGame applyMove(ChessGame g, ChessMove m) {
  final board=_applyRaw(g.board,m);
  var (wK,wQ,bK,bQ)=(g.wK,g.wQ,g.bK,g.bQ);
  var (epR,epC)=(-1,-1);
  final piece=g.board[m.r][m.c].abs();
  if(piece==K){if(g.turn==1){wK=wQ=false;}else{bK=bQ=false;}}
  if(piece==R){if(m.r==7&&m.c==7)wK=false;if(m.r==7&&m.c==0)wQ=false;if(m.r==0&&m.c==7)bK=false;if(m.r==0&&m.c==0)bQ=false;}
  if(m.doublePush){epR=(m.r+m.tr)~/2;epC=m.c;}
  final nextTurn=-g.turn;
  final legal=getLegalMoves(board,nextTurn,wK,wQ,bK,bQ,epR,epC);
  final status=legal.isEmpty?(_inCheck(board,nextTurn)?ChessStatus.checkmate:ChessStatus.stalemate):(_inCheck(board,nextTurn)?ChessStatus.check:ChessStatus.active);
  final half=(piece==P||g.board[m.tr][m.tc]!=0)?0:g.halfMove+1;
  final full=g.fullMove+(g.turn==-1?1:0);
  return ChessGame(board:board,turn:nextTurn,wK:wK,wQ:wQ,bK:bK,bQ:bQ,epR:epR,epC:epC,halfMove:half,fullMove:full,status:status,legalMoves:legal);
}

int _evaluate(List<List<int>> board) {
  var score=0;
  for(var r=0;r<8;r++) for(var c=0;c<8;c++){
    final p=board[r][c]; if(p==0) continue;
    final s=p>0?1:-1,a=p.abs(),idx=p>0?r*8+c:(7-r)*8+c;
    score+=s*((_VAL[a]??0)+(_PST[a]?[idx]??0));
  }
  return score;
}

int _mvvLva(List<List<int>> b, ChessMove m) {
  final v=b[m.tr][m.tc].abs(),att=b[m.r][m.c].abs();
  return v!=0?((_VAL[v]??0)*10-(_VAL[att]??0)):0;
}

int _minimax(ChessGame g, int depth, int alpha, int beta, int maximising) {
  if(depth==0||g.status==ChessStatus.checkmate||g.status==ChessStatus.stalemate) return _evaluate(g.board)*maximising;
  final moves=[...g.legalMoves]..sort((a,b)=>_mvvLva(g.board,b)-_mvvLva(g.board,a));
  var best=-(1<<30);
  for(final m in moves){
    final score=-_minimax(applyMove(g,m),depth-1,-beta,-alpha,-maximising);
    if(score>best)best=score; if(score>alpha)alpha=score; if(alpha>=beta)break;
  }
  return best;
}

ChessMove? getAiMove(ChessGame g) {
  if(g.legalMoves.isEmpty) return null;
  const depths=[0,1,1,2,2,2,3,3,4,4];
  final depth=depths[g.legalMoves.length<depths.length?g.legalMoves.length:depths.length-1];
  var bestScore=-(1<<30); var bestMove=g.legalMoves.first;
  for(final m in g.legalMoves){
    final score=-_minimax(applyMove(g,m),depth,-(1<<30),(1<<30),-g.turn);
    if(score>bestScore){bestScore=score;bestMove=m;}
  }
  return bestMove;
}
